d3a_interface/data_classes.py

```diff
@@ -0,0 +1,442 @@
+
+    def serializable_dict(self) -> Dict:
+        return {
+            "type": self.__class__.__name__,
+            "id": self.id,
+            "energy": self.energy,
+            "energy_rate": self.energy_rate,
+            "original_price": self.original_price,
+            "time": datetime_to_string_incl_seconds(self.time),
+            "attributes": self.attributes,
+            "requirements": self.requirements
+        }
+
+    @classmethod
+    def from_json(cls, offer_or_bid, current_time=None) -> Union["Offer", "Bid"]:
+        offer_bid_dict = json.loads(offer_or_bid)
+        object_type = offer_bid_dict.pop("type")
+
+        offer_bid_dict.pop("energy_rate", None)
+
+        if object_type == "Offer":
+            offer_bid_dict["time"] = current_time
+            return Offer(**offer_bid_dict)
+        if object_type == "Bid":
+            return Bid(**offer_bid_dict)
+
+
+class Offer(BaseBidOffer):
+    def __init__(self, id: str, time: datetime, price: float,
+                 energy: float, seller: str, original_price: Optional[float] = None,
+                 seller_origin: str = None, seller_origin_id: str = None,
+                 seller_id: str = None,
+                 attributes: Dict = None,
+                 requirements: List[Dict] = None):
+        super().__init__(id=id, time=time, price=price, energy=energy,
+                         original_price=original_price,
+                         attributes=attributes, requirements=requirements)
+        self.seller = seller
+        self.seller_origin = seller_origin
+        self.seller_origin_id = seller_origin_id
+        self.seller_id = seller_id
+
+    def __hash__(self) -> int:
+        return hash(self.id)
+
+    def __repr__(self) -> str:
+        return ("<Offer('{s.id!s:.6s}', '{s.energy} kWh@{s.price}', '{s.seller} {rate}'>"
+                .format(s=self, rate=self.energy_rate))
+
+    def __str__(self) -> str:
+        return ("{{{s.id!s:.6s}}} [origin: {s.seller_origin}] "
+                "[{s.seller}]: {s.energy} kWh @ {s.price} @ {rate}"
+                .format(s=self, rate=self.energy_rate))
+
+    def serializable_dict(self) -> Dict:
+        return {**super().serializable_dict(),
+                "seller": self.seller,
+                "seller_origin": self.seller_origin,
+                "seller_origin_id": self.seller_origin_id,
+                "seller_id": self.seller_id,
+                }
+
+    def __eq__(self, other) -> bool:
+        return (self.id == other.id and
+                self.price == other.price and
+                self.original_price == other.original_price and
+                self.energy == other.energy and
+                self.seller == other.seller and
+                self.seller_origin_id == other.seller_origin_id and
+                self.attributes == other.attributes and
+                self.requirements == other.requirements)
+
+    def csv_values(self) -> Tuple:
+        rate = round(self.energy_rate, 4)
```

Comment:
rounding is done in many places with same decimal places, better to replace it with a common rounding constant variable

Reply:
They are used 3 times with a value of 4 for the energy_rate, however I do not think that we should add a constant for them (at least not a global constant ) we may add a constant at the top of the file but The way I see it, this is not critical and would restrict us from having different decimal places for the existing different round calls
Anyway I could add the constant if you believe it's worth it

Reply:
We have a similar [round_floats_for_ui](https://github.com/gridsingularity/d3a-interface/blob/master/d3a_interface/utils.py#L302). So, could use it here too, off-course its rounding upto 3 decimal points.

Anyways non-blocking